docs/sys_modules_mock.py

```diff
@@ -0,0 +1,307 @@
+    "langchain.docstore.document",
+    "evidently.report",
+    "transformers.tokenization_utils_base",
+    "langchain.embeddings",
+    "great_expectations.profile.user_configurable_profiler",
+    "evidently.test_suite",
+    "deepchecks.vision.suites",
+    "google.auth.compute_engine",
+    "google.api_core",
+    "google.auth.credentials",
+    "google.oauth2.credentials",
+    "s3fs",
+    "gcsfs",
+    "boto3",
+    "adlfs",
+    "azure.keyvault",
+    "hvac",
+    "botocore",
+    "azure.keyvault.secrets",
+    "hvac.exceptions",
+    "botocore.exceptions",
+    "botocore.client",
+    "botocore.signers",
+]
+
+
+class DocsMocker(MagicMock):
+    _DOCS_BUILDING_MODE: bool = True
+
+    def __init__(self, *args, **kwargs):
+        super(DocsMocker, self).__init__(*args, **kwargs)
+        self.__getitem__ = DocsMocker._getitem
+        self._full_name: str = self._extract_mock_name()
+        self.__name__ = self._full_name.split(".")[-1]
+        self.__module__ = ".".join(self._full_name.split(".")[:-1])
+        self.__version__ = "1.0.0"
+        self.__mro_entries__ = lambda _: (self._factory(),)
+
+    def _factory(self, name: str = None):
+        if name:
+            class _(self.__class__):
+                __module__ = ".".join(name.split(".")[:-1])
+                __qualname__ = name.split(".")[-1]
+            return _(name=name)
+        else:
+            class _(self.__class__):
+                __module__ = self.__module__
+                __qualname__ = self.__name__
+
+            return _
+
+    @staticmethod
+    def _getitem(self: "DocsMocker", item):
+        return self._factory()(name=f"{self._full_name}[{item}]")
+
+    def __repr__(self) -> str:
+        return self._full_name
+
+
+for mocker in mockers:
+    sys.modules[mocker] = DocsMocker(name=mocker)
+
+from zenml.materializers.base_materializer import BaseMaterializer
+
+BaseMaterializer._DOCS_BUILDING_MODE = True
+
+if __name__ == "__main__":
+    import os
+    import importlib
+    from zenml.logger import get_logger
+
+    logger = get_logger(__name__)
+
+    modules_to_add = []
+    empty_dirs = []
+    meaningful_files = [".py", ".tf", ".yaml", ".yml"]
+    for root, subdirs, files in os.walk("src/zenml"):
+        if not root.endswith("__pycache__"):
+            root = root[len("src/") :]
+
+            no_meaningful_files = len(subdirs) == 0
+            for file in files:
+                for mf in meaningful_files:
+                    if file.endswith(mf):
+                        no_meaningful_files = False
+                        break
+                if file.endswith(".py") and not file == "__init__.py":
+                    module_name = root.replace("/", ".") + "." + file[:-3]
+                    try:
+                        importlib.import_module(module_name)
+                    except ModuleNotFoundError as e:
+                        msg: str = e.args[0]
+                        module = msg[msg.find("'") + 1 :]
```

Comment:
What's the +1 doing here? @coderabbitai  any ideas?

Maybe add a comment just to make it a bit clearer?

Reply:
Find function return a pos of a symbol, e.g. in `"str!payload".find("!")==4` and I need `payload`, so have to do `[4+1:]`

Reply:
I will add a comment